internal/controller/promotion/argocd.go

```diff
@@ -334,38 +345,68 @@ func (a *argoCDMechanism) mustPerformUpdate(
 		return "", true, errors.New("operation completed without a sync result")
 	}
 
+	// Check if the desired source(s) were applied.
+	if len(update.SourceUpdates) > 0 {
+		if (desiredSource != nil && !desiredSource.Equals(&status.SyncResult.Source)) ||
+			!desiredSources.Equals(status.SyncResult.Sources) {
+			// The operation did not result in the desired source(s) being applied.
+			// We should attempt to retry the operation.
+			return "", true, fmt.Errorf(
+				"operation result source does not match desired source",
+			)
+		}
+	}
+
 	// Check if the desired revision was applied.
-	if desiredRevision, err := libargocd.GetDesiredRevision(
+	desiredRevisions, err := libargocd.GetDesiredRevisions(
 		ctx,
 		a.kargoClient,
 		stage,
 		update,
 		app,
 		freightCol.References(),
-	); err != nil {
+	)
+	appLogger.Debug("Operation completed, checking applied revision.", "desiredRevisions", desiredRevisions)
+
+	if err != nil {
+		appLogger.Error(err, "Failed to determine desired revision, will need to retry.")
 		return "", true, fmt.Errorf("error determining desired revision: %w", err)
-	} else if desiredRevision != "" && status.SyncResult.Revision != desiredRevision {
+	}
+
+	if desiredRevisions == nil {
```

Comment:
It probably covers your bases better here if you check for `len(desiredRevisions) == 0` because that presumes less about the implementation of the `GetDesiredRevisions()` function. You know now that it returns a nil in certain cases, but if someone refactored that function in the future and ever returned an empty, but non-nil slice, your current check here would not be sufficient.